lib/http-bridge/src/Cardano/Wallet/HttpBridge.hs

```diff
@@ -0,0 +1,156 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: Apache-2.0
+--
+-- This module provides the main wallet server function for the
+-- cardano-http-bridge backend.
+
+module Cardano.Wallet.HttpBridge
+    ( serveWallet
+    ) where
+
+import Prelude
+
+import Cardano.BM.Trace
+    ( Trace, appendName, logAlert, logInfo )
+import Cardano.Launcher
+    ( ProcessHasExited (..), installSignalHandlers )
+import Cardano.Wallet
+    ( WalletLayer )
+import Cardano.Wallet.Api.Server
+    ( Listen (..) )
+import Cardano.Wallet.DaedalusIPC
+    ( daedalusIPC )
+import Cardano.Wallet.DB
+    ( DBLayer )
+import Cardano.Wallet.HttpBridge.Compatibility
+    ( HttpBridge )
+import Cardano.Wallet.HttpBridge.Environment
+    ( KnownNetwork (..) )
+import Cardano.Wallet.HttpBridge.Network
+    ( ErrStartup (..), HttpBridgeBackend (..) )
+import Cardano.Wallet.HttpBridge.Primitive.Types
+    ( Tx )
+import Cardano.Wallet.Network
+    ( NetworkLayer (..) )
+import Cardano.Wallet.Network.Ports
+    ( PortNumber )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( KeyToAddress )
+import Cardano.Wallet.Primitive.AddressDerivation.Sequential
+    ( SeqKey )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState )
+import Cardano.Wallet.Primitive.Types
+    ( Block )
+import Control.Concurrent
+    ( threadDelay )
+import Control.Concurrent.Async
+    ( race_ )
+import Control.Monad
+    ( forever )
+import Data.Function
+    ( (&) )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( ToText (..) )
+import Network.Wai.Handler.Warp
+    ( setBeforeMainLoop )
+import System.Exit
+    ( ExitCode (..) )
+
+import qualified Cardano.BM.Configuration.Model as CM
+import qualified Cardano.Wallet as Wallet
+import qualified Cardano.Wallet.Api.Server as Server
+import qualified Cardano.Wallet.DB.Sqlite as Sqlite
+import qualified Cardano.Wallet.HttpBridge.Network as HttpBridge
+import qualified Cardano.Wallet.HttpBridge.Transaction as HttpBridge
+import qualified Data.Text as T
+import qualified Network.Wai.Handler.Warp as Warp
+
+-- | The @cardano-wallet-http-bridge@ main function.
+serveWallet
+    :: forall t k n s a. (t ~ HttpBridge n, s ~ SeqState t, k ~ SeqKey)
+    => (KeyToAddress t k, KnownNetwork n)
+    => (CM.Configuration, Trace IO Text)
+    -- ^ Logging config.
+    -> Maybe FilePath
+    -- ^ Database file.
+    -> Listen
+    -- ^ HTTP API Server port.
+    -> HttpBridgeBackend
+    -- ^ Whether and how to launch or use the node backend.
+    -> Maybe (PortNumber -> NetworkLayer IO Tx (Block Tx) -> Sqlite.SqliteContext -> DBLayer IO s t k -> WalletLayer s t k -> IO a)
```

Comment:
This `SqliteContext` argument looks off / awkward to me here.

Reply:
Yeah unfortunately I can't get rid of it yet because this callback is used for tests and some of the tests want SqliteContext.